File: apps/macos/Sources/Library/WallpaperLibraryService.swift
import AVFoundation
import Foundation
import UniformTypeIdentifiers
import os

/// Imports, validates, and removes local wallpapers.
///
/// Imported files are copied into app-controlled storage. That keeps a
/// wallpaper working after the source is moved, renamed, or unplugged with an
/// external drive, and avoids security-scoped bookmark bookkeeping.
/// `@unchecked` only because `FileManager` carries no `Sendable` annotation;
/// the file operations used here are documented as thread-safe.
nonisolated struct WallpaperLibraryService: @unchecked Sendable {
    /// Container types accepted at import time.
    static let supportedContentTypes: [UTType] = {
        var types: [UTType] = [.mpeg4Movie, .quickTimeMovie, .gif]
        if let m4v = UTType("com.apple.m4v-video") {
            types.append(m4v)
        }
        return types
    }()

    static let supportedExtensions: Set<String> = ["mp4", "mov", "m4v", "gif"]

    /// What a converted GIF is stored as. Every other format is kept as it came.
    static let transcodedFileExtension = "mov"

    let storage: WallpaperStorage
    private let fileManager: FileManager

    init(storage: WallpaperStorage, fileManager: FileManager = .default) {
        self.storage = storage
        self.fileManager = fileManager
    }

    /// Validates `sourceURL`, copies it into storage, and returns the new
    /// record. `existing` is used to skip re-importing the same file twice.
    func importVideo(at sourceURL: URL, existing: [WallpaperRecord]) async throws -> WallpaperRecord {
        let name = sourceURL.lastPathComponent
        let needsConversion = GIFTranscoder.isGIF(sourceURL)

        guard Self.supportedExtensions.contains(sourceURL.pathExtension.lowercased()) else {
            throw WallpaperError.notPlayable(name: name)
        }

        // Validate before writing so an unusable file never enters storage. A
        // GIF is validated by the transcoder below, which has to open it anyway.
        let sourceMetadata = needsConversion
            ? nil
            : try await VideoMetadataReader.read(url: sourceURL, fileManager: fileManager)
        let sourceSize = fileSize(of: sourceURL)

        if let duplicate = existing.first(where: { isDuplicate($0, of: sourceURL, size: sourceSize) }) {
            Log.library.info("Skipping duplicate import of \(name, privacy: .public)")
            return duplicate
        }

        let id = UUID()
        let fileExtension = needsConversion
            ? Self.transcodedFileExtension
            : sourceURL.pathExtension.lowercased()
        let fileName = "\(id.uuidString).\(fileExtension)"
        let relativePath = storage.relativePath(directory: WallpaperStorage.videosDirectoryName, fileName: fileName)
        let destination = storage.url(forRelativePath: relativePath)

        do {
            try storage.prepareDirectories(fileManager: fileManager)
            if !needsConversion {
                try fileManager.copyItem(at: sourceURL, to: destination)
            }
        } catch {
            throw WallpaperError.importFailed(name: name, reason: error.localizedDescription)
        }

        // A converted GIF is measured from the movie that was just written, not
        // from the source, so the record describes what actually plays.
        let metadata: VideoMetadata
        if let sourceMetadata {
            metadata = sourceMetadata
        } else {
            metadata = try await GIFTranscoder.transcode(gifAt: sourceURL, to: destination)
        }

        let thumbnailPath = await makeImage(
            for: id,
            videoURL: destination,
            duration: metadata.duration,
            directory: WallpaperStorage.thumbnailsDirectoryName,
            maximumPixelWidth: ThumbnailGenerator.thumbnailPixelWidth
        )
        let stillPath = await makeImage(
            for: id,
            videoURL: destination,
            duration: metadata.duration,
            directory: WallpaperStorage.stillsDirectoryName,
            maximumPixelWidth: ThumbnailGenerator.stillPixelWidth
        )

        return WallpaperRecord(
            id: id,
            name: sourceURL.deletingPathExtension().lastPathComponent,
            relativePath: relativePath,
            thumbnailRelativePath: thumbnailPath,
            stillRelativePath: stillPath,
            dateImported: Date(),
            duration: metadata.duration,
            width: metadata.pixelSize.map { Int($0.width) },
            height: metadata.pixelSize.map { Int($0.height) },
            // The source's size, not the imported copy's: for a plain copy they
            // are the same, and for a converted GIF only the source's size can
            // recognise the same file being dropped again.
            fileSize: sourceSize,
            sourceFileName: name
        )
    }

    /// Deletes the imported copy and both generated images. Missing files are
    /// ignored: removing a record must always succeed from the user's point of
    /// view.
    func delete(_ record: WallpaperRecord) {
        let urls = [
            storage.videoURL(for: record),
            storage.thumbnailURL(for: record),
            storage.stillURL(for: record)
        ]
        for url in urls.compactMap({ $0 }) {
            try? fileManager.removeItem(at: url)
        }
    }

    /// Generates the full-resolution still for entries imported before stills
    /// existed, and returns the records that gained one so the caller can
    /// persist them. Entries whose video has gone are skipped.
    func backfillStills(for records: [WallpaperRecord]) async -> [WallpaperRecord] {
        var updated: [WallpaperRecord] = []

        for record in records where record.stillRelativePath == nil {
            guard fileExists(for: record) else { continue }

            let path = await makeImage(
                for: record.id,
                videoURL: storage.videoURL(for: record),
                duration: record.duration,
                directory: WallpaperStorage.stillsDirectoryName,
                maximumPixelWidth: ThumbnailGenerator.stillPixelWidth
            )
            guard let path else { continue }

            var record = record
            record.stillRelativePath = path
            updated.append(record)
        }

        return updated
    }

    func fileExists(for record: WallpaperRecord) -> Bool {
        fileManager.fileExists(atPath: storage.videoURL(for: record).path)
    }

    /// Removes stored files that no record points at, e.g. after a crash
    /// between the copy and the configuration save.
    func removeOrphanedFiles(keeping records: [WallpaperRecord]) {
        let keep = Set(records.flatMap { record in
            [record.relativePath, record.thumbnailRelativePath, record.stillRelativePath]
                .compactMap { $0 }
                .map { ($0 as NSString).lastPathComponent }
        })

        for directory in [storage.videosDirectory, storage.thumbnailsDirectory, storage.stillsDirectory] {
            let contents = (try? fileManager.contentsOfDirectory(atPath: directory.path)) ?? []
            for file in contents where !keep.contains(file) {
                try? fileManager.removeItem(at: directory.appendingPathComponent(file))
            }
        }
    }

    // MARK: - Private

    private func makeImage(
        for id: UUID,
        videoURL: URL,
        duration: Double,
        directory: String,
        maximumPixelWidth: CGFloat
    ) async -> String? {
        let relativePath = storage.relativePath(
            directory: directory,
            fileName: "\(id.uuidString).png"
        )
        let destination = storage.url(forRelativePath: relativePath)
        let generated = await ThumbnailGenerator.generate(
            from: videoURL,
            to: destination,
            duration: duration,
            maximumPixelWidth: maximumPixelWidth
        )
        return generated ? relativePath : nil
    }

    /// A repeat import is recognised by identical source file name and byte
    /// size. Records imported before renaming existed have no stored source
    /// name, so those fall back to the display name they were created with.
    private func isDuplicate(_ record: WallpaperRecord, of sourceURL: URL, size: Int64) -> Bool {
        let matchesName = record.sourceFileName.map { $0 == sourceURL.lastPathComponent }
            ?? (record.name == sourceURL.deletingPathExtension().lastPathComponent)

        return matchesName
            && record.fileSize == size
            && size > 0
            && fileExists(for: record)
    }

    private func fileSize(of url: URL) -> Int64 {
        let attributes = try? fileManager.attributesOfItem(atPath: url.path)
        return (attributes?[.size] as? NSNumber)?.int64Value ?? 0
    }
}

File: apps/windows/src/Wallpache.App/Library/WallpaperLibraryService.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Threading.Tasks;
using Wallpache.App.Support;

namespace Wallpache.App.Library;

/// <summary>
/// Imports, validates, and removes local wallpapers.
///
/// Imported files are copied into app-controlled storage. That keeps a
/// wallpaper working after the source is moved, renamed, or unplugged with an
/// external drive.
/// </summary>
public sealed class WallpaperLibraryService
{
    /// <summary>Container extensions accepted at import time.</summary>
    public static readonly string[] SupportedExtensions = [".mp4", ".mov", ".m4v"];

    public WallpaperLibraryService(WallpaperStorage storage)
    {
        Storage = storage;
    }

    public WallpaperStorage Storage { get; }

    /// <summary>
    /// Validates <paramref name="sourcePath"/>, copies it into storage, and
    /// returns the new record. <paramref name="existing"/> is used to skip
    /// re-importing the same file twice.
    /// </summary>
    public async Task<WallpaperRecord> ImportAsync(string sourcePath, IReadOnlyList<WallpaperRecord> existing)
    {
        var name = Path.GetFileName(sourcePath);
        var extension = Path.GetExtension(sourcePath).ToLowerInvariant();

        if (!SupportedExtensions.Contains(extension))
        {
            throw WallpaperException.NotPlayable(name);
        }

        // Validate before copying so an unusable file never enters storage.
        var metadata = await VideoMetadataReader.ReadAsync(sourcePath);
        var sourceSize = FileSize(sourcePath);

        var duplicate = existing.FirstOrDefault(record => IsDuplicate(record, sourcePath, sourceSize));
        if (duplicate is not null)
        {
            Log.Library.Info($"Skipping duplicate import of {name}");
            return duplicate;
        }

        var id = Guid.NewGuid();
        var relativePath = WallpaperStorage.RelativePath(
            WallpaperStorage.VideosDirectoryName,
            $"{id:D}{extension}");
        var destination = Storage.PathForRelative(relativePath);

        try
        {
            Storage.PrepareDirectories();
            File.Copy(sourcePath, destination, overwrite: false);
        }
        catch (Exception error)
        {
            throw WallpaperException.ImportFailed(name, error.Message);
        }

        var thumbnailPath = await MakeImageAsync(
            id,
            destination,
            metadata,
            WallpaperStorage.ThumbnailsDirectoryName,
            ThumbnailGenerator.ThumbnailPixelWidth);

        var stillPath = await MakeImageAsync(
            id,
            destination,
            metadata,
            WallpaperStorage.StillsDirectoryName,
            ThumbnailGenerator.StillPixelWidth);

        return new WallpaperRecord
        {
            Id = id,
            Name = Path.GetFileNameWithoutExtension(sourcePath),
            RelativePath = relativePath,
            ThumbnailRelativePath = thumbnailPath,
            StillRelativePath = stillPath,
            DateImported = DateTimeOffset.Now,
            Duration = metadata.Duration,
            Width = metadata.Width,
            Height = metadata.Height,
            FileSize = FileSize(destination),
            SourceFileName = name
        };
    }

    /// <summary>
    /// Deletes the imported copy and both generated images. Missing files are
    /// ignored: removing a record must always succeed from the user's point of
    /// view.
    /// </summary>
    public void Delete(WallpaperRecord record)
    {
        foreach (var path in new[] { Storage.VideoPath(record), Storage.ThumbnailPath(record), Storage.StillPath(record) })
        {
            if (path is null)
            {
                continue;
            }

            try
            {
                File.Delete(path);
            }
            catch (Exception error)
            {
                Log.Library.Error($"Could not delete {Path.GetFileName(path)}: {error.Message}");
            }
        }
    }

    /// <summary>
    /// Generates the full-resolution still for entries imported before stills
    /// existed, and returns the records that gained one so the caller can
    /// persist them. Entries whose video has gone are skipped.
    /// </summary>
    public async Task<IReadOnlyList<WallpaperRecord>> BackfillStillsAsync(IReadOnlyList<WallpaperRecord> records)
    {
        var updated = new List<WallpaperRecord>();

        foreach (var record in records.Where(record => record.StillRelativePath is null))
        {
            if (!FileExists(record))
            {
                continue;
            }

            var path = await MakeImageAsync(
                record.Id,
                Storage.VideoPath(record),
                new VideoMetadata(record.Duration, record.Width, record.Height),
                WallpaperStorage.StillsDirectoryName,
                ThumbnailGenerator.StillPixelWidth);

            if (path is null)
            {
                continue;
            }

            record.StillRelativePath = path;
            updated.Add(record);
        }

        return updated;
    }

    public bool FileExists(WallpaperRecord record) => File.Exists(Storage.VideoPath(record));

    /// <summary>
    /// Removes stored files that no record points at, e.g. after a crash between
    /// the copy and the configuration save.
    /// </summary>
    public void RemoveOrphanedFiles(IReadOnlyList<WallpaperRecord> records)
    {
        var keep = new HashSet<string>(
            records
                .SelectMany(record => new[] { record.RelativePath, record.ThumbnailRelativePath, record.StillRelativePath })
                .Where(path => !string.IsNullOrEmpty(path))
                .Select(path => Path.GetFileName(path!)),
            StringComparer.OrdinalIgnoreCase);

        foreach (var directory in new[] { Storage.VideosDirectory, Storage.ThumbnailsDirectory, Storage.StillsDirectory })
        {
            if (!Directory.Exists(directory))
            {
                continue;
            }

            foreach (var file in Directory.EnumerateFiles(directory))
            {
                if (keep.Contains(Path.GetFileName(file)))
                {
                    continue;
                }

                try
                {
                    File.Delete(file);
                }
                catch (Exception error)
                {
                    Log.Library.Error($"Could not remove orphan {Path.GetFileName(file)}: {error.Message}");
                }
            }
        }
    }

    // MARK: - Private

    private async Task<string?> MakeImageAsync(
        Guid id,
        string videoPath,
        VideoMetadata metadata,
        string directory,
        int maximumPixelWidth)
    {
        var relativePath = WallpaperStorage.RelativePath(directory, $"{id:D}.png");
        var destination = Storage.PathForRelative(relativePath);

        var generated = await ThumbnailGenerator.GenerateAsync(
            videoPath,
            destination,
            metadata.Duration,
            metadata.Width,
            metadata.Height,
            maximumPixelWidth);

        return generated ? relativePath : null;
    }

    /// <summary>
    /// A repeat import is recognised by identical source file name and byte
    /// size. Records imported before renaming existed have no stored source
    /// name, so those fall back to the display name they were created with.
    /// </summary>
    private bool IsDuplicate(WallpaperRecord record, string sourcePath, long size)
    {
        var matchesName = record.SourceFileName is not null
            ? string.Equals(record.SourceFileName, Path.GetFileName(sourcePath), StringComparison.OrdinalIgnoreCase)
            : string.Equals(record.Name, Path.GetFileNameWithoutExtension(sourcePath), StringComparison.OrdinalIgnoreCase);

        return matchesName && record.FileSize == size && size > 0 && FileExists(record);
    }

    private static long FileSize(string path)
    {
        try
        {
            return new FileInfo(path).Length;
        }
        catch (Exception)
        {
            return 0;
        }
    }
}
